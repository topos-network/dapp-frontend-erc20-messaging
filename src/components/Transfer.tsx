import { Button, Card, Form, Input, message, Modal, Space } from 'antd'
import { useMetaMask } from 'metamask-react'
import React from 'react'
import { Observable } from 'rxjs'

import TestId from '../util/testid'

interface Values {
  amount: number
  destinationAddress: string
}

interface TransferFormProps {
  visible: boolean
  setVisible: React.Dispatch<React.SetStateAction<boolean>>
  onSend: (values: Values) => Observable<string>
  onCancel: () => void
}

const TransferForm: React.FC<TransferFormProps> = ({
  visible,
  setVisible,
  onSend,
  onCancel,
}) => {
  const [disabled, setDisabled] = React.useState(false)
  const [form] = Form.useForm()

  return (
    <Modal
      visible={visible}
      title="Transfer an amount to an account"
      okText="Send"
      okButtonProps={{ disabled }}
      cancelText="Cancel"
      onCancel={onCancel}
      onOk={() => {
        form
          .validateFields()
          .then((values) => {
            message.loading({
              content: 'Sending the transfer',
              duration: 0,
              key: 'loader',
            })
            setDisabled(true)

            onSend(values).subscribe({
              next: () => {
                message.destroy('loader')
                message.success('The transfer was successfully processed')
                setVisible(false)
                setDisabled(false)
                form.resetFields()
              },
              error: (error) => {
                message.destroy('loader')
                message.error(error.message)
                setDisabled(false)
              },
            })
          })
          .catch((info) => {
            console.log('Validate Failed:', info)
          })
      }}
    >
      <Form
        form={form}
        layout="vertical"
        name="form_in_modal"
        initialValues={{
          modifier: 'public',
        }}
      >
        <Form.Item
          name="destinationAddress"
          label="Destination address"
          rules={[
            {
              required: true,
              message: 'Please input the destination address',
            },
          ]}
        >
          <Input data-testid={TestId.TRANSFER_ADDRESS_INPUT} />
        </Form.Item>
        <Form.Item
          name="amount"
          label="Amount"
          help="Unit"
          rules={[
            {
              required: true,
              message: 'Please input the amount to transfer',
            },
          ]}
        >
          <Input data-testid={TestId.TRANSFER_AMOUNT_INPUT} />
        </Form.Item>
      </Form>
    </Modal>
  )
}

export default () => {
  const { account } = useMetaMask()
  const [visible, setVisible] = React.useState(false)

  const onSend = React.useCallback(
    ({ amount, destinationAddress }: Values) => {
      return new Observable<string>()
    },
    [account]
  )

  return (
    <Card title="Transfer">
      <Space direction="vertical">
        <Button
          type="primary"
          data-testid={TestId.TRANSFER_BUTTON}
          onClick={() => {
            setVisible(true)
          }}
        >
          Make a transfer
        </Button>
        <TransferForm
          visible={visible}
          setVisible={setVisible}
          onSend={onSend}
          onCancel={() => {
            setVisible(false)
          }}
        />
      </Space>
    </Card>
  )
}
