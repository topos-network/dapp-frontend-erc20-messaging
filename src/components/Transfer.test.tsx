import React from 'react'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'

import TestId from '../util/testid'
import Transfer from './Transfer'

test('Should match snapshot', () => {
  const { asFragment } = render(<Transfer />)
  expect(asFragment()).toMatchSnapshot()
})

describe('About dialog', () => {
  beforeEach(async () => {
    render(<Transfer />)
    fireEvent.click(screen.getByTestId(TestId.TRANSFER_BUTTON))

    await waitFor(() => screen.getByText('Transfer an amount to an account'))
  })

  test('Should open dialog on button click', async () => {
    expect(screen.getByText('Transfer an amount to an account'))
  })

  test('Should close dialog on Cancel click', () => {
    fireEvent.click(screen.getByText('Cancel'))
    expect(!screen.getByText('Transfer an amount to an account'))
  })
})
