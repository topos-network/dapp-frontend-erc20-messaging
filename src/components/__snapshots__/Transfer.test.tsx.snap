// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="ant-card ant-card-bordered"
  >
    <div
      class="ant-card-head"
    >
      <div
        class="ant-card-head-wrapper"
      >
        <div
          class="ant-card-head-title"
        >
          Transfer
        </div>
      </div>
    </div>
    <div
      class="ant-card-body"
    >
      <div
        class="ant-space ant-space-vertical"
      >
        <div
          class="ant-space-item"
          style="margin-bottom: 8px;"
        >
          <button
            class="ant-btn ant-btn-primary"
            data-testid="7"
            type="button"
          >
            <span>
              Make a transfer
            </span>
          </button>
        </div>
        <div
          class="ant-space-item"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
